### PR TITLE
Update C# IsNameOfContext Extension

### LINKED ISSUE #1439: SyntaxTreeExtensions.IsNameOfContext is missing some cases, which causes rename to introduce errors
As a concrete example, rename "local" to "field" in this C# code:
```C#
class Program
{
    int field;

    static void Main(string[] args)
    {
        // Rename "local" to "field"
        int local;

        var x = nameof(field); // Correctly expands to Program.field
        nameof(field).ToString(); // Should also expand to Program.field
    }
}
```

**Expected**: 
```C#
class Program
{
    int field;

    static void Main(string[] args)
    {
        // Rename "local" to "field"
        int field;

        var x = nameof(Program.field); // Correctly expands to Program.field
        nameof(Program.field).ToString(); // Should also expand to Program.field
    }
}
```
**Actual**: 
```C#
class Program
{
    int field;

    static void Main(string[] args)
    {
        // Rename "local" to "field"
        int field;

        var x = nameof(Program.field); // Correctly expands to Program.field
        nameof(this.field).ToString(); // Should also expand to Program.field
    }
}
```

### PR DESCRIPTION
**Pull request by dpoeschl**
Fixes #1439

C#'s version of IsNameOfContext was written against a previous compiler
implementation of nameof in which the expression. Now, we only need to
check whether the nearest containing invocation expression is an
unqualified invocation of "nameof" that doesn't bind to anything.

### CODE REVIEW

File: src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs [hunk 1]
```diff
@@ -2117,71 +2117,18 @@ public static bool IsLabelContext(this SyntaxTree syntaxTree, int position, Canc
...
 
-                // Could have been parsed as a qualified name.
-                if (token.Parent.IsKind(SyntaxKind.QualifiedName))
-                {
-                    var parentQualifiedName = token.Parent;
-                    while (parentQualifiedName.IsParentKind(SyntaxKind.QualifiedName))
-                    {
-                        parentQualifiedName = parentQualifiedName.Parent;
-                    }
+            var invocationExpression = token.GetAncestor(a => a.IsKind(SyntaxKind.InvocationExpression)) as InvocationExpressionSyntax;
```
Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: Using `GetAncestor` here will walk up through lambda's into containing syntax that has `nameof`.  Is that okay? E.g. `nameof(a => { return 3; });` will be true inside the lambda body.

File: src/EditorFeatures/Test2/Rename/RenameEngineTests.CSharpConflicts.vb [hunk 2]
```diff
@@ -2963,7 +2964,7 @@ class C
 {
     void F(int [|$$z|])
     {
-        string x = nameof({|ref:zoo|});
```
Comment (Pilchie) [inline src/EditorFeatures/Test2/Rename/RenameEngineTests.CSharpConflicts.vb]: Why these changes?

**Comment by Pilchie** [comment]
Missing words in the description?

File: src/EditorFeatures/Test2/Rename/RenameEngineTests.CSharpConflicts.vb (hunk 2 above)
Comment (dpoeschl) [inline src/EditorFeatures/Test2/Rename/RenameEngineTests.CSharpConflicts.vb]: They exercise the change (and match the repro in the bug). I originally wrote this test this way and that's how I found this bug originally. I can make the test separate if needed,

File: src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs (hunk 1 above)
Comment (dpoeschl) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: I really didn't think this would matter for nameof. You can't even legally put expressions like that in a nameof, so if that code exists and nameof doesn't bind, you're either in broken code because you haven't created a nameof method yet or you're in broken code because you haven't realized that your subexpression is illegal within the nameof.

Comment (Pilchie) [inline src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs]: Yep, I realize that, but what is the user experience if those are treated like `nameof`?  Does it guide users down the wrong path.

File: src/EditorFeatures/Test2/Rename/RenameEngineTests.CSharpConflicts.vb (hunk 2 above)
Comment (Pilchie) [inline src/EditorFeatures/Test2/Rename/RenameEngineTests.CSharpConflicts.vb]: It's fine with me, was just curious about the reason for the change.

**Comment by dpoeschl** [comment]
This isn't the right fix. Closing until a proper fix is available.
